### PR TITLE
Add StochasticBlockModel type (SBM) and tests

### PR DESCRIPTION
**Pull request by jpfairbanks**
This PR takes the place of #226 with fewer commits.

### CODE REVIEW

File: src/randgraphs_sbm.jl [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    StochasticBlockModel(sizes, B)
+end
+
+function StochasticBlockModel(internalp::Vector{Float64}, externalp::Float64, sizes::Vector{Int})
+    B = sbmaffinity(internalp, externalp, sizes)
+    return StochasticBlockModel(sizes, B)
+end
+
+
+biclique = ones(2,2) - eye(2)
```
Comment (sbromberger) [inline src/randgraphs_sbm.jl]: const here?

**Comment by jpfairbanks** [comment]
GTG once tests pass. LGTM :ship:

File: src/randgraphs_sbm.jl [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+"""construct the affinity matrix for a near bipartite SBM.
+between is the affinity between the two parts of each bipartite community
+intra is the probability of an edge within the parts of the partitions.
+
+This is a specific type of SBM with k/2 blocks each with two halves.
+Each half is connected as a random bipartite graph with probability `intra`
+The blocks are connected with probability `between`.
+"""
+function nearbipartiteaffinity(sizes::Vector{Int}, between::Float64, intra::Float64)
+    numblocks = round(Int, length(sizes)/2)
```
Comment (sbromberger) [inline src/randgraphs_sbm.jl]: div or ÷ ?

File: src/randgraphs_sbm.jl [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+        produce( rand(1:n), rand(1:n) )
+    end
+end
+
+
+"""Take an infinite sample from the sbm.
+Pass to graph(edgestream, numedges) to get a Graph object.
+"""
+function sample(sbm::StochasticBlockModel)
+    pairs = @task random_pair(sbm.n)
```
Comment (sbromberger) [inline src/randgraphs_sbm.jl]: This is very clever.

File: src/randgraphs_sbm.jl [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        end
+        p = sbm.affinities[sbm.nodemap[i], sbm.nodemap[j]]
+        if rand() < p
+            produce(i, j)
+        end
+    end
+end
+
+"""convert a stream of edges produced by sample into a graph"""
+function graph(edgestream, sizes, numedges::Int)
```
Comment (sbromberger) [inline src/randgraphs_sbm.jl]: Not to bikeshed, but perhaps we should change the name of this function to `graph_from_stream` or something?

File: src/randgraphs_sbm.jl [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+    g = Graph(sum(sizes))
+    # println(g)
+    count = 1
+    for (i,j) in edgestream
+        # print("$count, $i,$j\n")
+        if !has_edge(g,i,j)
+            add_edge!(g,Edge(i,j))
```
Comment (sbromberger) [inline src/randgraphs_sbm.jl]: You might be able to do this more efficiently when we implement #246 :)

Also, you can just do `add_edge!(g, i, j)`.

File: src/randgraphs_sbm.jl [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            count += 1
+        end
+        if count >= numedges
```
Comment (sbromberger) [inline src/randgraphs_sbm.jl]: `count >= numedges && break`

File: src/randgraphs_sbm.jl [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    # @show Q'Q# < 1e-6
+    return (Q'A)*(Q)
+end
+
+
+function blockcounts(sbm::StochasticBlockModel, g::SimpleGraph)
+    return blockcounts(sbm, adjacency_matrix(g))
+end
+
+function blockfractions(sbm::StochasticBlockModel, g)
```
Comment (sbromberger) [inline src/randgraphs_sbm.jl]: can we parameterize g here? Looks like it can be `Union{SimpleGraph, AbstractMatrix}`?

File: test/sbm.jl [hunk 8]
```diff
@@ -0,0 +1,58 @@
+using Base.Test
+using LightGraphs
+import LightGraphs: nearbipartiteSBM, sample, StochasticBlockModel, graph, blockfractions, blockcounts
+
```
Comment (sbromberger) [inline test/sbm.jl]: You shouldn't need these if the functions are exported....

**Comment by sbromberger** [comment]
OK, go ahead and scream at me, but I think this should be added to randgraphs.jl. We don't need a separate file.

File: src/randgraphs_sbm.jl (hunk 6 above)
Comment (jpfairbanks) [inline src/randgraphs_sbm.jl]: :heavy_check_mark:

File: src/randgraphs_sbm.jl (hunk 2 above)
Comment (jpfairbanks) [inline src/randgraphs_sbm.jl]: :heavy_check_mark:

File: src/randgraphs_sbm.jl (hunk 4 above)
Comment (jpfairbanks) [inline src/randgraphs_sbm.jl]: :heavy_check_mark:

File: src/randgraphs_sbm.jl (hunk 7 above)
Comment (jpfairbanks) [inline src/randgraphs_sbm.jl]: :heavy_check_mark:

**Comment by jpfairbanks** [comment]
You can put it into randgraphs.jl if you want. I am not going to scream. I'll just let you do it :smile:
Line comments have been addressed.

The task doesn't stop, you could go back to it and get more edges if you want them. The reason for this design was to sample some edges, compute something, then sample some more edges and update the computation. You can see how feature evolve as the graph becomes denser.

**Comment by sbromberger** [comment]
If the task doesn't stop, does this mean that we've got another julia process hanging out just waiting for someone to consume random numbers?

**Comment by sbromberger** [comment]
and PS: I'll move it into randgraphs :)

**Comment by sbromberger** [comment]
See #261.

**Pull request merged**